dids/diddht/internal/dns/did.go

```diff
@@ -0,0 +1,278 @@
+		rootPropsSerialized = append(rootPropsSerialized, prop)
+	}
+
+	id := strings.TrimPrefix(d.ID, "did:dht:")
+	resource, err := newResource(fmt.Sprintf("_did.%s.", id), strings.Join(rootPropsSerialized, ";"))
+	if err != nil {
+		return nil, err
+	}
+	msg.Answers = append(msg.Answers, resource)
+
+	// add verification methods to dns message
+	for _, v := range d.VerificationMethod {
+		vm := v
+		// look for the key after the # in the verification method ID
+		key, ok := vmIDToK[v.ID]
+		if !ok {
+			// TODO handle error
+			continue
+		}
+		buf, err := MarshalVerificationMethod(&vm)
+		if err != nil {
+			return nil, err
+		}
+
+		resource, err := newResource(fmt.Sprintf("_%s._did.", key), buf)
+
+		if err != nil {
+			return nil, err
+		}
+
+		msg.Answers = append(msg.Answers, resource)
+	}
+
+	// add services to dns message
+	for _, s := range d.Service {
+		key, ok := sToK[s.ID]
+		if !ok {
+			// TODO handle error
+			continue
+		}
+		if err := MarshalService(key, s, &msg); err != nil {
+			return nil, err
+		}
+	}
+
+	msgByes, err := msg.Pack()
+	if err != nil {
+		return nil, err
+	}
+
+	return msgByes, nil
+}
+
+// UnmarshalDIDDocument unpacks the TXT DNS resource records and returns a DID document
+func UnmarshalDIDDocument(payload []byte) (*didcore.Document, error) {
+	decoder, err := parseDNSDID(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	doc, err := decoder.DIDDocument()
+	if err != nil {
+		return nil, err
+	}
+
+	return doc, nil
+}
+
+// MarshalVerificationMethod packs a verification method into a TXT DNS resource record and adds to the DNS message Answers
+func MarshalVerificationMethod(vm *didcore.VerificationMethod) (string, error) {
+	keyBytes, err := dsa.PublicKeyToBytes(*vm.PublicKeyJwk)
+	if err != nil {
+		return "", err
+	}
+
+	algID, err := dsa.AlgorithmID(vm.PublicKeyJwk)
+	if err != nil {
+		return "", err
+	}
+	t, ok := algToDhtIndex[algID]
+	if !ok {
+		return "", errors.New("unsupported algorithm")
+	}
+
+	props := []string{
+		"id=" + vm.ID,
+		"t=" + t,
+		"k=" + base64.RawURLEncoding.EncodeToString(keyBytes),
+	}
+
+	if len(vm.Controller) > 0 {
+		props = append(props, "c="+vm.Controller)
+	}
+
+	dhtEncodedVM := strings.Join(props, ";")
+
+	return dhtEncodedVM, nil
+
+}
+
+// MarshalService packs a service into a TXT DNS resource record and adds to the DNS message Answers
+func MarshalService(dhtDNSkey string, s *didcore.Service, msg *dnsmessage.Message) error {
+	rawData := fmt.Sprintf("id=%s;t=%s;se=%s", s.ID, s.Type, s.ServiceEndpoint)
+
+	resource, err := newResource(fmt.Sprintf("_%s._did.", dhtDNSkey), rawData)
+	if err != nil {
+		return err
+	}
+	msg.Answers = append(msg.Answers, resource)
+	return nil
+}
+
+// UnmarshalVerificationMethod unpacks the TXT DNS resource encoded verification method
+func UnmarshalVerificationMethod(data string, vm *didcore.VerificationMethod) error {
+	propertyMap, err := parseTXTRecordData(data)
+	if err != nil {
+		return err
+	}
+
+	vm.Type = "JsonWebKey2020"
```

Comment:
this must be `JsonWebKey`

Reply:
just for DHT or in general? `JsonWebKey2020` is used in other areas of the repo too

Reply:
eventually everywhere, for now definitely DHT https://github.com/TBD54566975/web5-spec/issues/73